demos/defender/defender_demo/metrics_collector.c

```diff
@@ -0,0 +1,381 @@
+
+            LogDebug( ( "File: %s, Line: %u, Content: %s.",
+                        pProcFile,
+                        lineNumber,
+                        &( lineBuffer[ 0 ] ) ) );
+
+            /* Skip the first line as it is a header. */
+            if( lineNumber <= 1 )
+            {
+                continue;
+            }
+
+            /* Parse the output. */
+            filledVariables = sscanf( &( lineBuffer[ 0 ] ),
+                                      "%*[^:]: %*8x:%4x %*8x:%*4x %2x",
+                                      &( localPort ),
+                                      &( connectionStatus ) );
+
+            /* sscanf should fill all the 2 variables successfully. */
+            if( filledVariables != 2 )
+            {
+                LogError( ( "Failed to parse %s.", &( lineBuffer[ 0 ] ) ) );
+                status = MetricsCollectorParsingFailed;
+                break;
+            }
+
+            if( connectionStatus == CONNECTION_STATUS_LISTEN )
+            {
+                if( pOutPortsArray != NULL )
+                {
+                    pOutPortsArray[ numOpenPorts ] = ( uint16_t ) localPort;
+                    numOpenPorts++;
+
+                    /* Break if the output array is full. */
+                    if( portsArrayLength == numOpenPorts )
+                    {
+                        break;
+                    }
+                }
+                else
+                {
+                    numOpenPorts++;
+                }
+            }
+        }
+    }
+
+    if( status == MetricsCollectorSuccess )
+    {
+        *pOutNumOpenPorts = numOpenPorts;
+    }
+
+    if( fileHandle != NULL )
+    {
+        fclose( fileHandle );
+    }
+
+    return status;
+}
+/*-----------------------------------------------------------*/
+
+MetricsCollectorStatus_t GetNetworkStats( NetworkStats_t * pOutNetworkStats )
+{
+    MetricsCollectorStatus_t status = MetricsCollectorSuccess;
+    FILE * fileHandle = NULL;
+    uint32_t lineNumber = 0, filledVariables;
+    uint32_t bytesReceived, bytesSent, packetsReceived, packetsSent;
+    char lineBuffer[ MAX_LINE_LENGTH ];
+
+    if( pOutNetworkStats == NULL )
+    {
+        LogError( ( "Invalid parameter. pOutNetworkStats: %p.", pOutNetworkStats ) );
+        status = MetricsCollectorBadParameter;
+    }
+
+    if( status == MetricsCollectorSuccess )
+    {
+        fileHandle = fopen( "/proc/net/dev", "r" );
+
+        if( fileHandle == NULL )
+        {
+            LogError( ( "Failed to open /proc/net/dev." ) );
+            status = MetricsCollectorFileOpenFailed;
+        }
+    }
+
+    if( status == MetricsCollectorSuccess )
+    {
+        /* Start with everything as zero. */
+        memset( pOutNetworkStats, 0, sizeof( NetworkStats_t ) );
+
+        while( fgets( &( lineBuffer[ 0 ] ), MAX_LINE_LENGTH, fileHandle ) != NULL )
+        {
+            lineNumber++;
+
+            LogDebug( ( "File: /proc/net/dev, Line: %u, Content: %s.",
```

Comment:
`/proc/net` filenames exist in quite a few places here.  You might consider making some `#define` values for these.

Reply:
I am not sure that would be any helpful.